modulus/datapipes/gnn/stokes_dataset.py

```diff
@@ -0,0 +1,306 @@
+
+        if self.num_samples > self.length:
+            raise ValueError(
+                f"Number of available {self.split} dataset entries "
+                f"({self.length}) is less than the number of samples "
+                f"({self.num_samples})"
+            )
+
+        self.graphs = []
+        for i in range(self.length):
+            # create the dgl graph
+            file_path = self.data_list[i]
+            polydata = read_vtp_file(file_path)
+            graph = self._create_dgl_graph(polydata, outvar_keys, dtype=torch.int32)
+            self.graphs.append(graph)
+
+        self.graphs = self.add_edge_features()
+
+        if self.split == "train":
+            self.node_stats = self._get_node_stats(keys=normalize_keys)
+            self.edge_stats = self._get_edge_stats()
+        else:
+            self.node_stats = load_json("node_stats.json")
+            self.edge_stats = load_json("edge_stats.json")
+
+        self.graphs = self.normalize_node()
+        self.graphs = self.normalize_edge()
+
+    def __getitem__(self, idx):
+        graph = self.graphs[idx]
+        return graph
+
+    def __len__(self):
+        return self.length
+
+    def add_edge_features(self):
+        """
+        adds relative displacement & displacement norm as edge features
+        """
+        for i in range(len(self.graphs)):
+            pos = self.graphs[i].ndata["pos"]
+            row, col = self.graphs[i].edges()
+            disp = torch.tensor(pos[row.long()] - pos[col.long()])
+            disp_norm = torch.linalg.norm(disp, dim=-1, keepdim=True)
+            self.graphs[i].edata["x"] = torch.cat((disp, disp_norm), dim=-1)
+
+        return self.graphs
+
+    def normalize_node(self):
+        """normalizes node features"""
+        invar_keys = set(
+            [
+                key.replace("_mean", "").replace("_std", "")
+                for key in self.node_stats.keys()
+            ]
+        )
+        for i in range(len(self.graphs)):
+            for key in invar_keys:
+                self.graphs[i].ndata[key] = (
+                    self.graphs[i].ndata[key] - self.node_stats[key + "_mean"]
+                ) / self.node_stats[key + "_std"]
+
+            self.graphs[i].ndata["x"] = torch.cat(
+                [self.graphs[i].ndata[key] for key in self.input_keys], dim=-1
+            )
+            self.graphs[i].ndata["y"] = torch.cat(
+                [self.graphs[i].ndata[key] for key in self.output_keys], dim=-1
+            )
+        return self.graphs
+
+    def normalize_edge(self):
+        """normalizes a tensor"""
+        for i in range(len(self.graphs)):
+            self.graphs[i].edata["x"] = (
+                self.graphs[i].edata["x"] - self.edge_stats["edge_mean"]
+            ) / self.edge_stats["edge_std"]
+
+        return self.graphs
+
+    @staticmethod
+    def denormalize(invar, mu, std):
+        """denormalizes a tensor"""
+        # assert invar.size()[-1] == mu.size()[-1]
```

Comment:
Are these comments needed?

Reply:
Thanks for pointing this out. No, this is not necessary and I have removed it.